
# Post \#67136061 [Link](https://stackoverflow.com/questions/67136061/)

## TypeError: list indices must be integers or slices, not str - While accessing element of list

**Vote**: 1 (672/702) **Views**: 15024 (357/702) 

**Internal ID** \#1-3-293

Created at 2021-04-17 08:11:31

Tags: `python` `python-3.x`

----------

#### Metadata:

Area: `Data Science & Machine Learning`

Language: `python` (full parsed tag list: `python`)

----------

**Notepad**


----------

I'm new to Python. Here I'm trying to access `list` object in python which is throwing following error.
> TypeError: list indices must be integers or slices, not str

---


`print(a)`

```
['{"id":9,"name":"Foo"}']
```

`print(type(a))`

```
<class 'list'>
```

`print(a['id'])`

```
TypeError: list indices must be integers or slices, not str
```

`print(a[0]['id'])`

```
TypeError: string indices must be integers
```

`print(a[0])`

```
{"id":9,"name":"Foo"}
```


---



### Expected Output


```
9
```



----------
        
## GPT4 Attempt \#0

The error you're encountering is because you're trying to access list elements using a string as an index. In Python, list indices must be integers or slices, not strings.

In your case, the list `a` contains a single string, which is a JSON formatted string. To access the 'id' value, you need to convert this JSON string into a Python dictionary first. You can do this using the `json.loads()` function from the `json` module.

Here's how you can do it:

```python
import json

a = ['{"id":9,"name":"Foo"}']

# Convert the JSON string into a dictionary
dict_a = json.loads(a[0])

# Now you can access the 'id' value
print(dict_a['id'])
```

This will output:

```
9
```

This code works by first importing the `json` module, then using the `json.loads()` function to convert the JSON string into a Python dictionary. After that, you can access the 'id' value just like you would in a normal dictionary.

--------

## GPT4 Attempt \#1

The error you're encountering is because the list `a` contains a string that is formatted like a dictionary, but it's still a string. Therefore, you can't access it like a dictionary.

To solve this, you need to convert the string to a dictionary using the `json.loads()` function from the `json` module. Here's how you can do it:

```python
import json

a = ['{"id":9,"name":"Foo"}']

# Convert the string to a dictionary
dict_a = json.loads(a[0])

# Now you can access it like a dictionary
print(dict_a['id'])
```

This will output:

```
9
```

The `json.loads()` function converts a JSON formatted string into a Python object, in this case a dictionary.

--------

## GPT4 Attempt \#2

The error you're encountering is because you're trying to access a list with a string index, which Python doesn't allow. List indices must be integers or slices, not strings.

In your case, the list `a` contains a single string that is formatted like a JSON object. To access the 'id' value, you need to convert that string into a dictionary first. You can do this using the `json` module in Python.

Here's how you can do it:

```python
import json

a = ['{"id":9,"name":"Foo"}']
dict_a = json.loads(a[0])

print(dict_a['id'])
```

In this code, `json.loads(a[0])` converts the string at index 0 of list `a` into a dictionary. Then, `dict_a['id']` accesses the value of 'id' in the dictionary.

--------
